### PR TITLE
chore: Update arrow, arrow-flight, parquet, tonic, prost, etc

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/influxdata/influxdb_iox/pull/4358~

# Rationale
Pull in changes from arrow 12 including updates to prost and tonic and dependents

Closes https://github.com/influxdata/influxdb_iox/pull/4318
Closes https://github.com/influxdata/influxdb_iox/pull/4362
Closes https://github.com/influxdata/influxdb_iox/pull/4363
Closes https://github.com/influxdata/influxdb_iox/pull/4364

# Changes:
1. Update to https://github.com/apache/arrow-datafusion/commit/7548e961b58f75fc26fd4fb16c99377f6f468fc9
2. Upgrade to arrow / parquet / arrow-flight 12
3. Update heappy: https://github.com/mkmik/heappy/pull/38 (thanks @mkmik !)

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1840,27 +1849,27 @@ dependencies = [
...
+ "prost-build 0.10.1",
+ "prost-types 0.10.0",
  "thiserror",
  "tokio",
  "tokio-stream",
  "tokio-util 0.7.1",
- "tonic 0.6.2",
- "tonic-build",
- "tonic-reflection",
+ "tonic 0.7.1",
```
Comment (alamb) [inline Cargo.lock:1859]: Old version of tonic is still around due to heappy dependency I think -- PR to update https://github.com/mkmik/heappy/pull/38

**Comment by alamb** [comment]
@tustvold  notes that we might be able to get rid of the protobuf changes (and need for a new compiler https://github.com/influxdata/influxdb_iox/pull/4358) if we remove the use of `optional`

For the record, here is how the test failed without the flag: https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/18207/workflows/6694ac22-c688-4ba8-bb8e-5d502752eb8d/jobs/151158

```
Error: Custom { kind: Other, error: "protoc failed: influxdata/iox/management/v1/server_config.proto:5:1: warning: Import google/protobuf/any.proto is unused.\ninfluxdata/iox/management/v1/service.proto:6:1: warning: Import google/protobuf/timestamp.proto is unused.\ninfluxdata/iox/management/v1/service.proto:7:1: warning: Import google/protobuf/any.proto is unused.\ninfluxdata/platform/storage/storage_common.proto:12:1: warning: Import influxdata/platform/storage/source.proto is unused.\ninfluxdata/iox/ingester/v1/query.proto: This file contains proto3 optional fields, but --experimental_allow_proto3_optional was not set.\n" }
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by alamb** [comment]
Update from @tustvold  --

> FYI I had a play with removing optional... It's kind of a pain... Probably best just to use a newer protoc, Flight SQL is using explicit field presence as well so

**Comment by crepererum** [comment]
I think we should just use a newer compiler if needed. If Debian doesn't provide it, we could also use Ubuntu-based CI images. The "use the newest stuff" strategy is also what we do for Rust and basically all dependencies and IMHO is a huge win for developer experience.

**Review by tustvold — APPROVED** [review]
:tada:

File: Cargo.lock [hunk 2]
```diff
@@ -4337,15 +4282,16 @@ dependencies = [
...
- "inferno 0.11.1",
+ "inferno",
  "libc",
  "log",
  "nix",
  "once_cell",
  "parking_lot 0.12.0",
  "prost 0.9.0",
- "prost-build",
+ "prost-build 0.9.0",
```
Comment (tustvold) [inline Cargo.lock:4292]: :cry:

Comment (alamb) [reply Cargo.lock:4292]: Actually there are outstanding PRs already: https://github.com/tikv/pprof-rs/pull/113. I asked the maintainer if they could merge and release an upgrade https://github.com/tikv/pprof-rs/pull/113#issuecomment-1103803657

Comment (alamb) [reply Cargo.lock:4292]: A new release is right around the corner. https://github.com/tikv/pprof-rs/pull/118 ❤️  Thank you @YangKeao

Comment (alamb) [reply Cargo.lock:4292]: https://github.com/influxdata/influxdb_iox/pull/4374 👍

**Pull request merged**